### ISSUE TITLE
I/O module: relationship between Serializers/Deserializers and channel

### ISSUE THREAD

**Issue by jhh67** [chapel-lang/chapel#18504]
Issue #18499 proposes new Encoder/Decoder functionality. This issue addresses the relationship between the encoders/decoders and channels (for brevity I will use the term "encoder" to mean "encoder/decoder"). As described in #18499 the encoder has an associated channel so that all operations on the encoder result in I/O to/from the channel. As an example: 
``` chapel
record Employee {
	var name: string;
	var id: int;
}

proc Employee.encodeTo(encoder: borrowed Encoder) throws {
	encoder.encode(key="name", name);
	encoder.encode(key="id", id);
}

var encoder = new JSONEncoder(stdout);
var bob = new Employee("Bob", 1);
encoder.encode(bob);  // {"name": "Bob", "id": 1}
```
This example has the effect of writing `{"name": "Bob", "id": 1}` to the `stdout` channel and the channel is passed as one of the arguments to `JSONEncoder()`. The advantage of this approach is that it allows multiple encoders to be used on one channel. This would be useful if the file contains multiple formats, e.g. only some of the data are in JSON.

An alternative is to associate the encoder with the channel so that all I/O to/from the channel goes through the encoder. As an example
```chapel
record Employee {
	var name: string;
	var id: int;
}

proc Employee.encodeTo(encoder: borrowed Encoder) throws {
	encoder.encode(key="name", name);
	encoder.encode(key="id", id);
}

var writer = openfile.writer(encoder = new JSONEncoder());
var bob = new Employee("Bob", 1);
writer.write(bob); // {"name": "Bob", "id": 1}
```
The advantage of this approach is that it allows the same encoder instance to be used across multiple channels.

### Open Questions
* Should a channel have a default encoder,e.g. Should `write(myCustomRecord)` call ‘encodeTo’ with some kind of default encoder?
  * In other words, how does `read` / `write` on a channel interact with the Encoder / Decoder?
  * Probably need something like this for `writeln(myrecord)` to work
  * This makes sense for the second option above in which channels have associated encoders. But, it could work with the first option too, if `write` etc calls create a `DefaultEncoder` as needed in order to format a record. Or, a channel could have a `DefaultEncoder` only but this encoder cannot be replaced by another one.
* Could/should we deprecate `writeThis`/`readThis`/`readWriteThis`?
  * A type could specify which fields are to be encoded in `encodeTo`.
  * A type could use tertiary methods on a `DefaultEncoder` to customize output.

**Comment by mppf** [comment]
> So, if we were to say that the encoder/decoder held a reference to its reader/writer rather than vice-versa, it seems to me that this would just be one more level of indirection between the file and the user which could compound this confusion and hierarchy

This resonates with me.

> * In terms of object design or a block diagram, I think of the encoder/decoder as being a filter that controls the way data gets passed through a reader/writer rather than something to be read from / written to itself, so patterns like `var output = new writer(encoder=myEncoder);  output.write(myVar);` seem more natural to me than `var enc = new encoder(stream=myStream);  enc.encode(myVar);`

This one does too.

> I tend to think that having a file + reader/writer that only cares about a single encoding/decoding strategy for its lifetime is a common case and somehow having the channel store its reader/writer seems more aligned with that case than the opposite to me, though I realize that's just a matter of opinion (as are most of these points).

I think you meant having the channel store its encoder/decoder in the above, right?

I think my main concern with having the channel store the encoder/decoder is that it seems awkward to change it, if we support that at all. Occasionally I have worked with file formats that use a mix of text and binary (for example) which might be represented with different encoders (or at least it is a related situation). It's probably OK though to either:
 * save the result of encoding in the unusual format to a `bytes` (or something) and then write that
 * or create another channel to the same file (which is not always possible with a pipe or something) and work with that.
 
A related question is how we can allow the 1st case to work. With the `encoder.encode(myobject)` strategy, we still have `channel.write(mybytes)` available to us. But if the channel's encoder is chosen to be, say, JSON, and then we call `channel.write(mybytes)`, won't it necessarily write the bytes as a JSON string?

So that leads me to thinking that maybe we have `read()`/`readln()` and `write()`/`writeln()` always working through the Decoder/Encoder, but that there are some I/O calls that never do? For example we have `readbits` and `readbytes` and `writebits` and writeBytes` today. Presumably these will always ignore the Decoder/Encoder. Does that make sense? I think of the proposed `writeBigEndian` and similar functions as in the same category.

So, does it seem OK that we will have reader/writer methods that do I/O divided into two categories - low level operations that ignore the Encoder/Decoder - and higher level operations that work with it?

**Comment by mppf** [comment]
(For that matter, some file formats will specify the format of newlines, so `readln()` / `writeln()` might not actually work with the Encoder. With the `encoder.encode(mything)` strategy, this is not a problem. But what should a binary format Encoder (say, for the sake of argument, Google Protobuf Encoder) do with the "add a newline" part in `writeln` ? Should some encoders just throw if there's a `readln` / `writeln`?

**Comment by bradcray** [comment]
> I think you meant having the channel store its encoder/decoder in the above, right?

Yeah, thanks, I've fixed it.

> it seems awkward to change it, if we support that at all

In saying that, do you mean "awkward syntactically for the user" or "it raises awkward challenges in the implementation"?  I don't find either of the proposals for overriding a channel's default encoder/decoder particularly awkward (with a slight preference for the version that overwrites it for a given `write()` call):

> to have the ability to either change that encoder dynamically (e.g., `outstream.setEncoder(myEncoder)`, or else to override it for a specific read/write call (e.g., `outstream.write(x, y, z, encoder=myEncoder);`).

But I'm definitely not thinking a lot about the implementation...  Abstractly, for the second, I imagine "save the original encoder away, replace it with the new one, do the write, and then replace the original.  I can see that there could be a parallel safety issue lurking here, but given that we already make given read/write calls atomic-ish w.r.t. other read/write calls, that seems like it shouldn't be an issue?  If there weren't significant challenges, this seems nicer than requiring the user to open a second channel or manually create a bytes representing mixed format content (which I agree seem a bit heroic).

> there are some I/O calls that never do?
> ...
> So, does it seem OK that we will have reader/writer methods that do I/O divided into two categories - low level operations that ignore the Encoder/Decoder - and higher level operations that work with it?

I buy all that.

> Should some encoders just throw if there's a readln / writeln?

Either that, or document that they ignore the `ln` part?  (I imagine any encoder/decoder may encounter types or input strings that it can't handle and will have to throw on them, so throwing in this case doesn't seem that different.  But ignoring it may be more convenient for users who are accustomed to using `writeln()` most of the time, say (while also potentially permitting code reuse?).

**Comment by mppf** [comment]
Thanks for reminding me about `outstream.write(x, y, z, encoder=myEncoder);`. I agree that would neatly solve the problem of needing mixed encoding types.

Regarding the I/O calls that work with the encoder we have this proposal:

Works with encoder:
 * read, readln, write, writeln (where the `ln` part might be ignored or cause the encoder to throw if it can't do that)
 
Ignores encoder:
 * readbits, readbytes, writebits, writeBytes
 * writeBigEndian etc (to be added)
 
I have three follow-up questions about this.

First, what should happen if a read/write call uses multiple arguments? Should the Encoder/Decoder just assume that the data is concatenated? I would suppose that won't work with all file formats (e.g. imagine a file format where `""` means something different, we `write("a", "b")` and it emits `"a""b"` but `""` might have special meaning in the file format (after all `"""` does). I'm not sure how realistic this problem is, though). Alternatives are -- to consider the arguments to be bundled together into a tuple or to only allow read/write calls with non-default Encoder/Decoder with 1 argument.

Second, how does formatted I/O with `readf` / `writef` interact with the Encoder/Decoder? I think we have been expecting to deprecate the '%ht' and '%jt' formats. Should there still be `writef("%t", myrecord)` to "do whatever `write` would do? Is there any expectation that the format string will allow customization? It does today but I don't think this is reasonable to continue to support with Encoder/Decoder. E.g. we can have this today:

``` chapel
record R {
  var field: int;
}
writef("%xt\n", new R(10)); // prints (field = 0xa)
```

Third, one of the things I have been happy about with the Decoder design is that we pass the type in to the decode call (not a value). This seems a bit fraught if we are combining with read/write, especially with multiple arguments. In particular, I want the value decoded to be created and returned in the call - and not have to default initialize anything. This can work with `read(type t)` and if we add `read(out value: ?t)`. But, today, we can't simply implement a varargs `read(out args...)` function because some of the actual arguments are values (`ioLiteral` and `ioNewline`). Additionally, today, `read` on classes expects them to be allocated before the `read` call. This won't work with an `out` argument either. Now, maybe the changes here are coming up for `read` anyway (e.g. deprecating `ioLiteral` and `ioNewline` in favor of separate calls and always using `out` on the arguments). But if not, it seems it would be cleaner/less invasive to use a different call.

(All 3 of these questions are easily answered in the `encoder.encode()` approach -- so I'd characterize them as challenges with channel-has-encoder/decoder approach).

**Comment by mppf** [comment]
One thing that occurs to me now is that, as a starting point, we could add `channel.encode()` and `channel.decode()`. These would always use the encoder (and nothing else). Then, we can make some of the read/write routines call these as we have been discussing. I think this would be nice for a couple of reasons:
 * it allows us to separate the design questions about read/write from what we need to get started with implementing the Encoder/Decoder (so allows progress)
 * it allows somebody using a channel where they only want to work with the encoder or decoder to express that clearly (since arbitrary I/O in the middle of the encoded file can mess up a file format if done accidentally)
 * it makes it easier in the documentation to describe what is happening in I/O functions that do a mix of encoding and writing. For example, if we keep `read` / `write` with the signatures they have now, and we keep `ioLiteral` and `ioNewline`, we could say something like "`channel.write` will output all record arguments passed to it with `channel.encode`". Of course the alternative is to say something like "will use the Encoder attached to the channel" but that seems to be harder to explain. 

Even in the long term, it seems reasonable to me to have both `channel.encode`/`channel.decode` and have some `channel.write`/`channel.read` calls that end up calling `encode`/`decode`. Connected to that - I think we need to have `write(myRecord)` / `read(myRecord)` work with an Encoder / Decoder (and I think we agree about that) but the details there seem to have more to do with write/read than anything else.

**Comment by lydia-duncan** [comment]
> Should some encoders just throw if there's a readln / writeln?

While I think this is natural, we did back away from making `readln` and `writeln` throw earlier because we didn't want users to have to surround calls to something as basic as `writeln` with a try/catch statement.  This seems like a step back towards that and while it does seem justified, I think it's something we should consider carefully.

**Comment by mppf** [comment]
`channel.writeln` throws but the top-level `writeln` does not. I was talking about the `channel.writeln`.

**Comment by benharsh** [comment]
On this issue we had discussed adding some kind of optional "encoder" argument to the ".write" methods, but I don't think that's possible with our current implementation of varargs. Consider the following example:

```chpl
record X { }
record Y { }

proc foo(args...?k, encoder:? = new X()) {
  writeln(args);
  writeln("encoder: ", encoder.type:string);
}

foo(1,2,3);
foo(1,2,3, encoder=new Y());
```

This program prints:
```
(1, 2)
encoder: int(64)
(1, 2, 3)
encoder: Y
```

With our current implementation, varargs are not really "greedy" and the last argument of ``foo`` will always bind to the ``encoder`` formal unless we use a named argument. It's also worth pointing out that we compute vararg bindings/alignment *before* we instantiate formals. If that's sufficiently surprising or disappointing I can make an issue about whether we should make that change at a language level.

In the meantime, we could add a new method that would create a shallow alias of the channel with a separate encoder:
```chpl
// could use "encode" as an alternatively shorter method name?
myWriter.withEncoder(jsonEncoder).write("hello");
```

A couple other options:
```chpl
// Add an "encode" method that always accepts an encoder and otherwise acts like "write"
myWriter.encode(jsonEncoder, "hello");

// Add a format specifier to 'writef' that knows to look for an encoder - kinda weird
myWriter.writef("%i %v %s\n", 5, (jsonEncoder, "hello"), "goodbye"); // prints: 5 "hello" goodbye
```

**Comment by bradcray** [comment]
> With our current implementation, varargs are not really "greedy" and the last argument of foo will always bind to the encoder formal unless we use a named argument.

Huh, that's interesting...  I think it _might_ be worth asking the question (opening the issue) whether this is a bug or feature.  That said, I suspect the big challenge here is that all the args are so generic.  E.g., changing your example to a varargs of int doesn't work either, presumably because the encoder argument is also so generic.  My next thought was "let's make the encoder argument not be generic", but I guess a downside of using the record-based approach to encoders (and not having interfaces) is that that argument needs to remain completely generic as well?  Or should/could we have `record encoder {` that is used to wrap _all_ encoders (like domains and array records, but without so much sugar) and then make that argument be `encoder(?)`?

All that said, I think an only slightly lame way to get the effect of what you were trying to sketch out could be achieved by cloning the routine and leaning on "last resort":  https://tio.run/##hY3BCsIwEETv@YolpwRKQL21iL9hjyHdxmJMyiZSpPjtsdYSPAjedmb2zZiLHtHlTGgCdXCGGZ5sE@0q2EjBQB@C0GSjUup0lTAzgImGhM6vtmy@DC467PXdJUBvQock@ZKvTdreNHCnYwLCGCjxH/VV4f7sbG818IKo9BixjokGbz@j7N29q/bVYdHlLsDR4wStkLLJ@QU

but this has the mild downside of wanting to implement the no-encoder version by having it call the encoder-version, which gets back to the original problem.  But there are also ways to work this like having a helper routine that both call, where the encoder argument is moved to the first position..., e.g.:  https://tio.run/##bY/NagIxFIX3eYpDNmYgjKg7RXwNXZUwXmds00m4iUgRn31MVEJou7uHfOcn3WA82Wli6hwfsccNd/EWh6cQnl2Hk3MfA1lPrGjs3JFYw3Af2rbdfTW4CeDK50h2VPINYAupUWi5rmQbfzytQ@Tz2Deb2pxDX858NRuIaoL6VVmtGumKvWo0VHp/WpMX8zkc4/MSIrwJAXEgxIu3hFlmZrtXuum/DaQ1CWMKjqP8p7KM/9Nd/lh152CR/Qu91Kuky11ytnnzQSV4mh4

**Comment by benharsh** [comment]
My worry with something like the "last resort" approach is that it's not really something we'd recommend to users, but something they might want to know how to do if they see it actually working.

I've wondered if first-class functions could help us here, but I don't think we have a clear enough idea of how generics could or should work. Something like: ``fileWriter.write(args...?k, encoder: proc(ch: fileWriter, const x: ?t)) { ... }``

**Comment by bradcray** [comment]
Yeah, that's partially why I called it slightly lame.  But the "only" qualifier is because I think we should make the "last resort" feature user-facing somehow... just not sure how.

That said, the other reason I called it slightly lame is that it feels a little unfortunate to lean on it for this case.

Oh wait, I just remembered another way to potentially get this pattern.  Only, it... doesn't work?  Though I'm not sure why it wouldn't / doesn't, and it definitely feels more like a bug to me, or at least a case worthy of discussion, than the original case.  The idea is to put a dummy argument of a type that the user won't/can't pass into writeln in-between:

```chapel
proc foo(args...?k, dummy: nothing = none, encoder:? = new X()) {
  writeln(args);
  writeln("encoder: ", encoder.type:string);
}

foo(1,2,3);
foo(1,2,3, encoder=new Y());
```

but I get a bug where it's trying to pass '2' to 'dummy'.  Maybe this is due to the greedy behavior you mentioned above, where I had been thinking that that was only problematic in the original case because both args and encoder were completely generic.

That makes me think your FCF-based approach would likely hit the same issue (as would my previous suggestion that we use a single generic record type with a class guts s.t. we could constrain the argument to be `: encoder(?)` to break where the varargs matching occurs.

**Comment by benharsh** [comment]
I do think there's a bug of some sort here, but I still would have expected an error passing ``3`` to ``dummy``. Right now we compute the binding of actuals to formals before we consider the types. So there isn't logic that looks at ``dummy`` and says "oh, I can't pass that thing to dummy, so let's give it to the preceding varargs instead if possible".

Which means that my FCF idea wouldn't have worked either for the same reason.

**Comment by bradcray** [comment]
The reason that surprises me is that I would've thought that our logic was (or "would be") "have the varargs greedily gobble up all arguments until/unless we reach a named argument."  Or maybe what I mean is:
* match all named arguments first
* then positionally match remaining arguments
* if one of the positional arguments is varargs, have it gobble up all arguments
  * (or optionally "gobble up all arguments that match any type constraint it has", but you're suggesting that that's not the case)

Also, I think I know why I thought we supported it: Just because we'd tried to do it before, see `functions/varargs/issue-17188.chpl`

**Comment by bradcray** [comment]
Another test that looks related to this: `functions/diten/defaultValueArgAfterVararg.chpl`

**Comment by mppf** [comment]
I think it probably wouldn't be to hard to fix #17188 if we think the named-argument is the best design here. My own opinion is that `withEncoder` is a great tool that we would want either way. So my preference would be to proceed with that and then add the named-argument part later if `withEncoder` proves to be too awkward.

I think we should prioritize fixing #17188 if we think it has an impact on language stabilitiy.

> [edit: oops, meant to put that on https://github.com/chapel-lang/chapel/issues/17188, but don't want to do that now and potentially generate more mails for folks on these threads]

I edited your last response there to include it, which doesn't generate a mail, and gets it in the right place.

**Comment by lydia-duncan** [comment]
@benharsh - do you think this could be closed, or would you prefer to wait until the remaining serializer/deserializer tasks have completed?

**Comment by benharsh** [comment]
I'd like to keep this open for the time being, and will close it once I've written up some of the documentation that addresses the questions posed by the issue.

**Comment by benharsh** [comment]
I'm closing this issue now since its original purpose has been fulfilled and the relevant changes are in the 1.32 release. I've opened https://github.com/chapel-lang/chapel/issues/23528 to continue the discussion about having an optional argument for ``writeln`` that can take a serializer.
